### ISSUE TITLE
[Feature Request]: Comparing files solely by the file name has unpleasant drawbacks

### ISSUE THREAD

**Issue by MartinJDavis93** [sboesen/remotely-sync#40]
### What feature are you suggesting?

I remembered another not-so-pleasant feature. Currently, as far as I understand, file comparison is done only based on the file name. So, if you create a file on a device with the name File1, synchronize it across your devices, and then decide to rename it, for example, to File2, it won't actually be renamed to File2. Instead, the original File1 will be deleted (sent to the trash) because a file with that name no longer exists. File2 will be synchronized as a separate file, not connected to File1 in any way. And each subsequent renaming of the file will erase the previous file. So, if you later want to rename File2 to File3, you will already find File1 and File2 in the trash, which essentially are all the same; you just decided to rename them. And that's a problem. It would be more reasonable to compare files, for example, by the file hash or some other additional indicator besides the name, or assign a separate ID to each file, which would be stored in a separate file list. However, implementing such a change would not be an easy task.

### Ensure no sensitive information

- [X] I ensure that no sensitive information is submitted in the issue.

**Comment by sboesen** [comment]
Definitely a good feature request which will sadly require some large updates to the sync algorithm to support this.

For encrypted vaults I think we could do this by adding the hash for all synced files to the encrypted manifest. Otherwise we'd be fetching and re-hashing files just to compare locally, which would probably add up to be an expensive operation especially on mobile.

Might be worth making this an optional feature, which adds more sync complexity. Will have to think more about implementation details. Thanks for opening!

**Comment by MartinJDavis93** [comment]
> Definitely a good feature request which will sadly require some large updates to the sync algorithm to support this.
> 
> For encrypted vaults I think we could do this by adding the hash for all synced files to the encrypted manifest. Otherwise we'd be fetching and re-hashing files just to compare locally, which would probably add up to be an expensive operation especially on mobile.
> 
> Might be worth making this an optional feature, which adds more sync complexity. Will have to think more about implementation details. Thanks for opening!

Yes, this will completely change the logic of the entire synchronization. Moreover, depending on the implementation, it will take more time for synchronization. And a lot of time will be spent on such implementation, not to mention the nuances that may arise during the implementation process. Although it is important in principle, I think it's still worth postponing such implementation for the time being, focusing on more urgent needs. I just decided to outline this issue before I forget about it. Because over all these years of use, it hasn't been a big problem for me, more like an unpleasant peculiarity of work that you can tolerate and eventually don't even notice. I think many people haven't even paid attention to it.

**Comment by sboesen** [comment]
Agreed, it definitely can be an issue but it is largely avoidable. I think it's worth tracking, agreed, but due to the heavy lift might take some time to implement.

**Comment by MartinJDavis93** [comment]
> Agreed, it definitely can be an issue but it is largely avoidable. I think it's worth tracking, agreed, but due to the heavy lift might take some time to implement.

Yes, it's definitely a problem. I wouldn't have posted it otherwise. Moreover, depending on the synchronization service, it could potentially become a security issue. But, I reiterate, this will require additional investigation, and accordingly, I assume, a considerable investment of time for implementation and stabilization of this fix. If, after all the major fixes, you'll still have the energy and motivation to implement it, I'm all for it.

**Comment by MikeWi11s0n** [comment]
+1

**Comment by OzzmRT** [comment]
I tried Remotely Sync instead of Remotely Save, hoping that this bug was finally fixed, but unfortunately, it wasn't. I don't entirely agree with the author that it's just a "not-so-pleasant feature". It's a rather serious bug that leads to the loss of original files if you decide to rename them while Obsidian is open on another device. So, I hope you will take note of this and try to fix it.

**Comment by sboesen** [comment]
> Wouldn't this be fixable with something like: `this.registerEvent(app.on('rename', callback));`

Unfortunately we already [do this](https://github.com/sboesen/remotely-sync/blob/1dbdf872fbc3da4b05bc06ad3dd3676d0fb589c4/src/main.ts#L518). This only tracks rename events from *within* obsidian (which is tracked in a seperate issue #56). A real fix will need to be more specific to this issue, assigning files unique IDs like @MartinJDavis93 suggested is probably the most durable solution.

> It's a rather serious bug that leads to the loss of original files if you decide to rename them while Obsidian is open on another device.

@OzzmRT Are you losing files when renaming files? I wonder if you had two syncs going on at the same time. I feel like this is a separate bug from the issue described, as that won't affect all situations, but agreed this feature implementation will solve the issue. Probably not worth filing a separate issue, but understand the urgency.

**Comment by sboesen** [comment]
Closing in favor of #56.

**Comment by sboesen** [comment]
Reopening - the more I think about it neither #56 nor #70 will perfectly solve the issue as described originally. This requires its own fix (basically add a hash to #70 as an additional check before creating new files to handle renames across devices when syncing doesn't happen in the ideal path).
